src/slam_karto.cpp

```diff
@@ -68,6 +65,44 @@
 #include <utility>
 #include <vector>
 
+// Specialize some graph access methods for the graph publisher
+namespace locus_mapping_tools
+{
+
+template <>
+std::vector<locus_msgs::Edge> getEdges(const karto::MapperGraph& graph)
+{
+  auto edges = std::vector<locus_msgs::Edge>();
+  for (const auto& edge : graph.GetEdges())
+  {
+    auto edge_msg = locus_msgs::Edge();
+    edge_msg.node_ids[0] = edge->GetSource()->GetObject()->GetUniqueId();
+    edge_msg.node_ids[1] = edge->GetTarget()->GetObject()->GetUniqueId();
+    edges.push_back(edge_msg);
+  }
+  return edges;
+}
+
+template <>
+std::vector<locus_msgs::Node> getNodes(const karto::MapperGraph& graph)
+{
+  auto nodes = std::vector<locus_msgs::Node>();
+  for (const auto& [name, vertices] : graph.GetVertices())
```

Comment:
Is it possible to use boost, something like:
```
auto name_vertices = graph.GetVertices();
for_each (name_vertices.begin(), name_vertices.end(), boost::bind(karto::MapperGraph::GetVertices(),
boost::bind<karto::MapperGraph&>(&graph::value_type::second,_1)));

```

Reply:
There is a std::for_each now as well.

I did think about it, but it felt weird with the nested loops.
* I would need to define a lambda expression inside another lambda expression in order to handle the nesting. And although legal, that felt ugly. Either that, or I would need to implement some free functions to handle the conversions from karto to message types.
* I would have to capture and then modify the `nodes` variable. I try to avoid having my lambda functions have side effects like that. They are not necessarily obvious, and are easy to overlook while debugging.

I could use an STL algorithm for just the inner loop. That feels much more natural:
```
template <>
std::vector<locus_msgs::Node> getNodes(const karto::MapperGraph& graph)
{
  auto convert_vertex = [](const karto::Vertex* vertex)
  {
      auto node_msg = locus_msgs::Node();
      node_msg.id = vertex->GetObject()->GetUniqueId();
      node_msg.position.x = vertex->GetObject()->GetCorrectedPose().GetX();
      node_msg.position.y = vertex->GetObject()->GetCorrectedPose().GetY();
      return node;
  };

  auto convert_edge = [](const karto::Edge* edge)
  {
      auto edge_msg = locus_msgs::Edge();
      edge_msg.node_ids[0] = edge->GetSource()->GetObject()->GetUniqueId();
      edge_msg.node_ids[1] = edge->GetTarget()->GetObject()->GetUniqueId();
      return edge_msg;
  };

  auto nodes = std::vector<locus_msgs::Node>();
  for (const auto& [name, karto_vertices] : graph.GetVertices())
  {
    std::transform(karto_vertices.begin(), karto_vertices.second.end(), std::back_inserter(nodes), convert_vertex);
  }
  
  auto edges = std::vector<locus_msgs::Edge>();
  const auto& karto_edges = graph.GetEdges();
  std::transform(karto_edges.begin(), karto_edges.second.end(), std::back_inserter(edges), convert_edge);
```

Reply:
Yes you-re right STL offered `std::for_each` it slipped my mind.